packages/adapters/src/adapters/Cornerstone/Segmentation_4X.js

```diff
@@ -1541,3 +1660,109 @@ function getSegmentMetadata(multiframe, seriesInstanceUid) {
         data
     };
 }
+
+function readFromUnpackedChunks(chunks, offset, length) {
+    const mapping = getUnpackedOffsetAndLength(chunks, offset, length);
+
+    // If all the data is in one chunk, we can just slice that chunk
+    if (mapping.start.chunkIndex === mapping.end.chunkIndex) {
+        return new Uint8Array(
+            chunks[mapping.start.chunkIndex].buffer,
+            mapping.start.offset,
+            length
+        );
+    } else {
+        // If the data spans multiple chunks, we need to create a new Uint8Array and copy the data from each chunk
+        let result = new Uint8Array(length);
+        let resultOffset = 0;
+
+        for (
+            let i = mapping.start.chunkIndex;
+            i <= mapping.end.chunkIndex;
+            i++
+        ) {
+            let start =
+                i === mapping.start.chunkIndex ? mapping.start.offset : 0;
+            let end =
+                i === mapping.end.chunkIndex
+                    ? mapping.end.offset
+                    : chunks[i].length;
+
+            result.set(
+                new Uint8Array(chunks[i].buffer, start, end - start),
+                resultOffset
+            );
+            resultOffset += end - start;
+        }
+
+        return result;
+    }
+}
+
+function getUnpackedOffsetAndLength(chunks, offset, length) {
+    var totalBytes = chunks.reduce((total, chunk) => total + chunk.length, 0);
+
+    if (offset < 0 || offset + length > totalBytes) {
+        throw new Error("Offset and length out of bounds");
+    }
+
+    var startChunkIndex = 0;
+    var startOffsetInChunk = offset;
+
+    while (startOffsetInChunk >= chunks[startChunkIndex].length) {
+        startOffsetInChunk -= chunks[startChunkIndex].length;
+        startChunkIndex++;
+    }
+
+    var endChunkIndex = startChunkIndex;
+    var endOffsetInChunk = startOffsetInChunk + length;
+
+    while (endOffsetInChunk > chunks[endChunkIndex].length) {
+        endOffsetInChunk -= chunks[endChunkIndex].length;
+        endChunkIndex++;
+    }
+
+    return {
+        start: { chunkIndex: startChunkIndex, offset: startOffsetInChunk },
+        end: { chunkIndex: endChunkIndex, offset: endOffsetInChunk }
+    };
+}
+
+function calculateCentroid(imageIdIndexBufferIndex, multiframe) {
```

Comment:
It feels to me like some of these should be added to a helper/utils export rather than being directly included here, as they may well be useful in other segmentation generations in the future, and also that allows explaining what they are doing cleanly.

Reply:
it is really not a general utility as it is relying on imageIdIndexBufferIndex